UDF Style Updates

WilkinsKa1 opened the pull request:
## Description

• placeholder text on UDFs now displays as helper text
•  updating multiselect to use react-select
• updating oneof-select to be a regular select if options > 2, otherwise default to radio display. 
• adding password prop to make helper text conditional 

## Type of Change
- Non breaking change ('update', 'change')

## Note to Reviewers

Please check that multiselect and oneof select components are still good. Currently there aren't any one click apps that have a multiselect UDF, but some of the security stackscripts under community do- you can view the updated component there.

martinmckenna commented on src/features/StackScripts/UserDefinedFieldsPanel/FieldTypes/UserDefinedSelect.tsx:
```diff
@@ -48,41 +53,67 @@ class UserDefinedSelect extends React.Component<CombinedProps, State> {
+          {error && <Notice error text={error} spacingTop={8} />}
+          <InputLabel className={classes.radioGroupLabel}>
+            {field.label}
+            {!isOptional && ' *'}
```
<img width="384" alt="Screen Shot 2019-03-27 at 5 11 10 PM" src="https://user-images.githubusercontent.com/7387001/55112610-56f86c00-50b3-11e9-8eb9-baba64a3c56d.png">

Small nitpick, but text and password fields don't have a space between the work and asterisk

martinmckenna commented on src/features/StackScripts/UserDefinedFieldsPanel/FieldTypes/UserDefinedMultiSelect.tsx:
```diff
@@ -30,92 +24,46 @@ interface Props {
-    const newSelections = selectedValuesToArray.join(',');
-
-    updateFormState(field.name, newSelections);
+  handleChangeMulti = (valueMulti: Item[]) => {
```
this isn't updating the form state of the `LinodeCreateContainer`. It needs to invoke `props.updateFormState`. You can probably get rid of `this.state.valueMulti` altogether and just rely on `props.udf_data`

WilkinsKa1 commented:
Thanks @martinmckenna , I updated per your feedback.

martinmckenna commented:
I think the UDF multi select is still malfunctioning. After console logging the output of the UDF state, I get something like this

<img width="896" alt="Screen Shot 2019-04-01 at 8 54 45 AM" src="https://user-images.githubusercontent.com/7387001/55329191-ff178780-545b-11e9-807e-a67d99f3b9f5.png">

Notice how the object nested under the `allow` key is a duplicate of the parent object.

The `manyof` UDFs need to be send back as a comma separated list. So when you select an option from the dropdown, the full UDF payload should look something like

```js
{
  LOGRO_FREQ: "monthly"
  LOGRO_ROTA: "12"
  /** This is the manyof UDF - notice how it's a comma-seperated list */
  allow: "NTP Service: UDP 123,Mail Submission: TCP 587,SSL IMAP Server: TCP 993"
  icmplevel: "Ignore all ICMP"
  loglevel: "Nothing"
  sshd_group: "sshusers"
  sshd_passwordauth: "Yes"
  sshd_permitroot: "No"
  sshd_port: "22"
  sshd_protocol: "2"
  sshrange: "0/0"
  sudo_usergroup: "wheel"
}
```

Let me know if you still need help / further explanation about the old code and what exactly it was doing

WilkinsKa1 commented:
@alioso @martinmckenna Thanks for reviewing! 
Alban- I updated the error to have `spacingTop`.  I need to get clarification on where the error is coming from, doesn't seem to be defined in manager. The workaround for testing is to select another plan (ex. Nanode) and that should get you through a successful creation.

Marty- Thanks for providing screenshots, I updated so now `allow` will come back as a comma separated string.

martinmckenna commented on src/features/StackScripts/UserDefinedFieldsPanel/FieldTypes/UserDefinedMultiSelect.tsx:
```diff
@@ -30,92 +24,51 @@ interface Props {
   state: State = {
-    manyof: this.props.field.manyof!.split(',')
+    manyof: this.props.field.manyof!.split(','),
+    selectedOptions: null
```
One more minor thing - I think you can get rid of this `selectedOptions` state altogether. react select appears to handle the value internally

WilkinsKa1 replied:
@martinmckenna updated!

martinmckenna commented on src/features/StackScripts/UserDefinedFieldsPanel/FieldTypes/UserDefinedMultiSelect.tsx:
```diff
@@ -30,92 +24,51 @@ interface Props {
-      (udf_data[field.name] && udf_data[field.name].split(',')) || [];
+    const arrayToString = Array.prototype.map
+      .call(selectedOptions, (opt: Item) => opt.value)
+      .toString();
```
TIL that `Array.prototype.toString()` converts things to a comma-seperated list. who knew?
